### ISSUE TITLE
Support field projection inside of `Unaligned` (and other wrappers?)

### ISSUE THREAD

**Issue by joshlf** [google/zerocopy#196]
TODO: Fill in details here.

**Comment by SkiFire13** [comment]
> ```rust
>         unsafe {
>             $crate::$project_fn(
>                 $c.$borrow_method(),
>                 |outer| outer as $ptr_ty,
>                 |inner| ::core::ptr::$addr_of!((*inner) $($f)*),
>                 |field| field as $ptr_ty,
>             )
>         }
> ```

Note that this allowes `c` and `$($f)*` to do `unsafe` operations without actually using the `unsafe` keyword (because it's hidden in your macro)

**Comment by joshlf** [comment]
Good catch, thanks for that, @SkiFire13 ! Took a quick stab at fixing it and ran into inference issues (specifically when assigning `|inner| ...` to a variable before the `unsafe` block), but I've added this to the design doc; we'll obviously need to fix that before the design is finalized.

**Comment by SkiFire13** [comment]
> Took a quick stab at fixing it and ran into inference issues (specifically when assigning `|inner| ...` to a variable before the `unsafe` block)

A couple of ideas for this:

- you could use those expressions somewhere else, where they are never executed, but outside of an `unsafe` block so the compiler still checks them. For example inside an `if false { ...; unreachable!() }` block.

- you could try making an identity function that just specifies the bounds on the closure, helping type inference. Something like:

  ```
  fn help_outer_to_inner_inference<P, OuterToInner>(f: OuterToInner) -> OuterToInner
  where
      P: Projectable + ?Sized,
      OuterToInner: FnOnce(*const P) -> *const P::Inner
  {
      f
  }
  ```

  and likewise for the other closures.

- another approach could be moving the required use of `unsafe` somewhere else, for example by making the function safe, but having it take something like an `UnsafeToken`, which can only be created by using `unsafe`. This way it's as if the function is `unsafe`, but you're not required to wrap the whole call expression, including the parameters, in it.

**Comment by joshlf** [comment]
Found unsoundness in the way we're using `addr_of!`: https://users.rust-lang.org/t/how-to-use-addr-of-to-create-an-unaligned-pointer/98524

**Comment by joshlf** [comment]
> * another approach could be moving the required use of `unsafe` somewhere else, for example by making the function safe, but having it take something like an `UnsafeToken`, which can only be created by using `unsafe`. This way it's as if the function is `unsafe`, but you're not required to wrap the whole call expression, including the parameters, in it.

Ended up taking this approach, thanks for the suggestion, @SkiFire13!

**Comment by joshlf** [comment]
Another challenge we've just run into: The code as it's currently designed allows for an arbitrary sequence of field accesses, index operations, and slice operations. However, if a field is a reference type, this has the consequence of indirecting through the reference. This breaks the soundness of the `project!` macro. Consider, for example, `MaybeUninit<&'static u8>`. It would not be sound to field project this into `&'static MaybeUninit<u8>` - the original reference might be uninitialized!

We need to figure out a way to reject any projection through references.
